### ISSUE TITLE
Avoid LayoutStride when slicing the 2nd slowest dimension. 

### ISSUE THREAD

**Issue by bartgol** [kokkos/kokkos#3757]
Currently, in `Kokkos_ViewMapping.hpp`, there's logic to preserve LayoutRight/Left in a subview under certain conditions, if the output view has rank<=2. For instance, if `v` has rank 3 (dynamic) and LayoutRight, we'd have

```
auto A = Kokkos::ALL();
auto s1 = Kokkos::subview(v,1,A,A);  // Still LayoutRight: ViewOffset::m_stride is 0
auto s2 = Kokkos::subview(v,A,1,A);  // Still LayoutRight: ViewOffset::m_stride is v.extent(1)*v.extent(2)
auto s3 = Kokkos::subview(v,A,A,1);  // LayoutStride
```
However, if v has rank N, taking a N-1 slice of v would preserve LayoutRight only when slicing the 1st dimension. However, slicing the 2nd dimension would still allow to only store a single m_stride. However, when trying to subview, say, `View<int****,LayoutRight>` with args `(A,0,A,A)` yileds LayoutStride.

Is there a reason for this that I can't see? Or is it just that when the logic was implemented, the case of a N-1 slice over the 2nd dimension wasn't needed?

It seems to me that the only thing that would need to be changed is the logic [here](https://github.com/kokkos/kokkos/blob/1fb0c284d458c75370094921d9f202c287502325/core/src/impl/Kokkos_ViewMapping.hpp#L3692), to allow keeping the input layout if rank = SrcRank-1, and R1==0.

In theory, one could accommodate the most general case, which requires every sliced index (i.e., when the Arg in the subview is a single int) does not have more than 1 extent in the slower striding dims. E.g., a rank-8 could be subviewed as `subview(v,A,0,0,0,0,0,A,A)` or `subview(v,0,0,0,0,A,0,A,A)` and still require to store only a single `m_stride`. However, these cases are more difficult, and might require some mods in `SubviewExtents`. 

But it would be nice to subview a 4 or 5 dim view at the 2nd slowest index and still maintain the right/left layout of the input view. That is, allow the following:

```
auto A = ALL();
View<int********,LayoutRight> v("",10,10,10,10,10,10,10,10);
auto sR = subview(v,A,0,A,A,A,A,A,A); // Keep LayoutRight
View<int********,LayoutLeft> v("",10,10,10,10,10,10,10,10);
auto sL = subview(v,A,A,A,A,A,A,0,A); // Keep LayoutLeft
```
I could put up a small PR if this is acceptable.

**Comment by crtrott** [comment]
Uhm we can think about it but boy this is getting complicated.

**Comment by bartgol** [comment]
> Uhm we can think about it but boy this is getting complicated.

Yes, it does add more logic to the subview layout deduction. But maybe not _too_ much. Let take a quick shot at it.

**Comment by bartgol** [comment]
Ugh, it was more complicated than I originally thought. I have a branch, which I can share, if need be, but I'm not sure if it would ever help.

Anyhow, please, consider this issue a very low priority.
